src/OrleansRuntime/Scheduler/ShedulerExtensions.cs

```diff
@@ -7,64 +7,23 @@ internal static class SchedulerExtensions
     {
         internal static Task<T> QueueTask<T>(this OrleansTaskScheduler scheduler, Func<Task<T>> taskFunc, ISchedulingContext targetContext)
         {
-            var resolver = new TaskCompletionSource<T>();
-            Func<Task> asyncFunc =
-                async () =>
-                {
-                    try
-                    {
-                        T result = await taskFunc();
-                        resolver.TrySetResult(result);
-                    }
-                    catch (Exception exc)
-                    {
-                        resolver.TrySetException(exc);
-                    }
-                };
-
-            // it appears that it's not important that we fire-and-forget asyncFunc() because we wait on the
-            scheduler.QueueWorkItem(new ClosureWorkItem(() => asyncFunc().Ignore()), targetContext);
-            return resolver.Task;
+            var workItem = new AsyncClosureWorkItem<T>(taskFunc);
+            scheduler.QueueWorkItem(workItem, targetContext);
+            return workItem.Task;
         }
 
         internal static Task QueueTask(this OrleansTaskScheduler scheduler, Func<Task> taskFunc, ISchedulingContext targetContext)
         {
-            var resolver = new TaskCompletionSource<bool>();
-            Func<Task> asyncFunc =
-                async () =>
-                {
-                    try
-                    {
-                        await taskFunc();
-                        resolver.TrySetResult(true);
-                    }
-                    catch (Exception exc)
-                    {
-                        resolver.TrySetException(exc);
-                    }
-                };
-            scheduler.QueueWorkItem(new ClosureWorkItem(() => asyncFunc().Ignore()), targetContext);
-            return resolver.Task;
+            var workItem = new AsyncClosureWorkItem(taskFunc);
+            scheduler.QueueWorkItem(workItem, targetContext);
+            return workItem.Task;
         }
 
         internal static Task QueueNamedTask(this OrleansTaskScheduler scheduler, Func<Task> taskFunc, ISchedulingContext targetContext, string activityName = null)
         {
-            var resolver = new TaskCompletionSource<bool>();
-            Func<Task> asyncFunc =
-                async () =>
-                {
-                    try
-                    {
-                        await taskFunc();
-                        resolver.TrySetResult(true);
-                    }
-                    catch (Exception exc)
-                    {
-                        resolver.TrySetException(exc);
-                    }
-                };
-            scheduler.QueueWorkItem(new ClosureWorkItem(() => asyncFunc().Ignore(), () => activityName), targetContext);
```

Comment:
@gabikliot thanks for pointing that out. If you trace the execution, you can that there's one use of this method, and (eventually) it's from `GrainTimer` where it's always set to `Name` (a property on that class).

Digging further backwards, `GrainTimer.Name` is set to a non-constant value in a couple of cases:
* Reminders, where it's set to the reminder name.
* Streams (`PersistentStreamPullingAgent` and `PersistentStreamPullingManager`), where it's set to the value of `CurrentActivationContext.Name` + "Timer" << so this one allocates.

In this case, we're already allocating the string and holding onto it for a longer duration than the scheduled task.

It's possible I missed something, though.